### PR TITLE
WIP move module-var and -accessor synthesis out of mixin

### PR DESCRIPTION
**Pull request by retronym**

### CODE REVIEW

File: src/compiler/scala/tools/nsc/transform/Mixin.scala [hunk 1]
```diff
@@ -1082,18 +1060,7 @@ abstract class Mixin extends InfoTransform with ast.TreeDSL {
...
-
-            // !!! TODO - unravel the enormous duplication between this code and
-            // eliminateModuleDefs in RefChecks.
-            val rhs          = gen.newModule(sym, vsym.tpe)
-            val assignAndRet = gen.mkAssignAndReturn(vsym, rhs)
-            val attrThis     = gen.mkAttributedThis(clazz)
-            val rhs1         = mkInnerClassAccessorDoubleChecked(attrThis, assignAndRet, sym, List())
-
-            addDefDef(sym, rhs1)
+            // Moved to Refchecks
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/Mixin.scala]: Noice!

**Comment by retronym** [comment]
Factored out some duplication between `elimModuleDef` and `elimMixinModuleDef` in the latest commit.

Testing: https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-validate-main/374/console

**Comment by retronym** [comment]
Take 2 with an early sprinkling of `makeNotPrivate`: https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-validate-main/375/console

**Comment by retronym** [comment]
:cry: 

https://scala-ci.typesafe.com/job/scala-2.12.x-validate-publish-core/370/console

```
extras
[quick.partest-extras] Compiling 11 files to /home/jenkins/workspace/scala-2.12.x-validate-publish-core/build/quick/classes/partest-extras
[quick.partest-extras] OpenJDK 64-Bit Server VM warning: ignoring option MaxPermSize=512M; support was removed in 8.0
[quick.partest-extras] OpenJDK 64-Bit Server VM warning: Using the ParNew young collector with the Serial old collector is deprecated and will likely be removed in a future release
[quick.partest-extras] /home/jenkins/workspace/scala-2.12.x-validate-publish-core/src/partest-extras/scala/tools/partest/Util.scala:17: error: macro implementation has incompatible shape:
[quick.partest-extras]  required: (c: scala.reflect.macros.blackbox.Context)(a: c.Expr[A]): c.Expr[A]
[quick.partest-extras]  or      : (c: scala.reflect.macros.blackbox.Context)(a: c.Tree): c.Tree
[quick.partest-extras]  found   : (c: scala.reflect.macros.blackbox.Context)(a: c.Expr[A]): c.Expr[A]
[quick.partest-extras] type mismatch for return type: c.Expr[?A] does not conform to c.Expr[A]
[quick.partest-extras]   def trace[A](a: A) = macro traceImpl[A]
[quick.partest-extras]                              ^
[quick.partest-extras] one error found
```

**Comment by retronym** [comment]
```
⚡cat sandbox/test.scala
import language.experimental._, reflect.macros.blackbox.Context

object Util {
  def trace[A](a: A) = macro traceImpl[A]
  def traceImpl[A: c.WeakTypeTag](c: Context)(a: c.Expr[A]): c.Expr[A] = ???
}

⚡ qscalac -uniqid -Ydebug -Ymacro-debug-verbose sandbox/test.scala
[running phase parser on test.scala]
[running phase namer on test.scala]
[running phase packageobjects on test.scala]
[running phase typer on test.scala]
typechecking macro def term macro trace#7671 at source-/Users/jason/code/scala/sandbox/test.scala,line-4,offset=85
generating macroImplSigs for: <method> <macro> <locked> def trace#7671[A#7673&0](a#15160: A) = traceImpl[A]
result is: MacroImplSig([A#7673&0](c: blackbox#7654.this.Context#7667)(a: c#22755.Expr#22426[A#7673&0])c#24023.Expr#22426[A#7673&0])
blackbox#7654.this.Context#7667 <: blackbox#7654.this.Context#7667?
true
Exprs#7742.this.Expr#10394[T#10399] <: Trees#8600.this.Tree#9271?
false
c#22755.Expr#22426[A#7673&0] <: c#22755.Expr#22426[?A]?
true
Exprs#7742.this.Expr#10394[T#10399] <: Trees#8600.this.Tree#9271?
false
c#22755.Expr#22426[?A] <: c#24023.Expr#22426[A#7673&0]?
  c#22755.type = c#24023.type?
  false
  c#22755.universe#22409.type = c#24023.universe#22409.type?
    c#22755.type = c#24023.type?
    false
  false
false
sandbox/test.scala:4: error: macro implementation has incompatible shape:
```

**Comment by retronym** [comment]
My change to mixin module accessors / vars in refchecks is currently stuck on an obscure error in the macro engine that only shows up after bootstrapping. I’m trying to hone in on the problem by using a non-boostrapped quick build (`ant -Dlocker.skip=1 quick.bin`), compiling `src/compiler` with that to a temp dir, then adding that temp dir to the JVM classpath of the quick compiler. That fails in the same way, so I’m about to get bash to delete the bootstrapped class files one by one until the problem disappears, then get onto some bytecode diffing...

// Time passes.

Found it. I was calling `makeNotPrivate` on the trait-defined module when mixing it into the class, so as to make the mixed-in accessor take on the matching, expanded name. But this only renamed the module + module class, but did not rename the companion class. This inhibited the “case apply” to “constructor call” transform in refchecks.

Not sure yet how that change led to the behaviour change in `Validators`. I'm going to try to disable that optimization in refchecks in a branch based on 2.12.x, and check if we run into the same error.

**Comment by retronym** [comment]
Hmmm, `Validators` is still broken after a bootstrap, the case class `apply` transform must have been a red herring.

**Comment by retronym** [comment]
Okay, I'm forgetting to memoize method-local modules, this fails:

```
object Test {
  def main(args: Array[String]): Unit = {
    object O
    val x = O
    val y = O
    assert(x eq y)
  }
}
```

**Comment by retronym** [comment]
https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-validate-main/377/console

**Comment by retronym** [comment]
A few checkfile updates later: https://scala-ci.typesafe.com/job/scala-2.12.x-validate-test/298/console

**Comment by adriaanm** [comment]
Many down, one to go!
```
fail - neg/t6666c.scala  [compilation failed]% scalac t6666c.scala
error: value <none> is not a member of <notype>
```

**Comment by retronym** [comment]
https://scala-ci.typesafe.com/job/scala-2.12.x-validate-main/409/console

**Comment by retronym** [comment]
```
  [partest] ok 1705 - run/t1987.scala                         
  [partest] ok 1706 - run/t6687.scala                         
  [partest] ok 1707 - run/duplicate-meth.scala                
  [partest] ok 1708 - run/t6052.scala                         
  [partest] ok 1709 - run/t0508.scala                         
  [partest] ok 1710 - run/type-currying.scala                 
Build timed out (after 150 minutes). Marking the build as aborted.
Build was aborted
Archiving artifacts
Notifying endpoint 'HTTP:http://scala-ci.typesafe.com:8888/jenkins'
Finished: ABORTED
```

**Comment by retronym** [comment]
```
⚡ jstack 22381
2015-08-31 16:12:09
Full thread dump Java HotSpot(TM) 64-Bit Server VM (25.51-b03 mixed mode):

"Attach Listener" #5316 daemon prio=9 os_prio=31 tid=0x00007f9b1b3a5800 nid=0x6bfb waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"pool-1-thread-5" #15 prio=5 os_prio=31 tid=0x00007f9b158bb000 nid=0x5d03 runnable [0x0000000125486000]
   java.lang.Thread.State: RUNNABLE
	at scala.collection.IndexedSeqLike$Elements.next(IndexedSeqLike.scala:65)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:453)
	at scala.collection.Iterator$class.foreach(Iterator.scala:781)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1267)
	at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:155)
	at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1267)
	at scala.collection.TraversableOnce$class.$div$colon(TraversableOnce.scala:149)
	at scala.collection.AbstractIterator.$div$colon(Iterator.scala:1267)
	at scala.collection.SetLike$class.$plus$plus(SetLike.scala:150)
	at scala.collection.AbstractSet.$plus$plus(Set.scala:47)
	at scala.tools.nsc.backend.jvm.analysis.AliasingFrame.valuesWithAliasId(AliasingFrame.scala:41)
	at scala.tools.nsc.backend.jvm.analysis.AliasingFrame.aliasesOf(AliasingFrame.scala:46)
	at scala.tools.nsc.backend.jvm.analysis.AliasingFrame.scala$tools$nsc$backend$jvm$analysis$AliasingFrame$$$anonfun$4(AliasingFrame.scala:227)
	at scala.tools.nsc.backend.jvm.analysis.AliasingFrame.merge(AliasingFrame.scala:226)
	at scala.tools.asm.tree.analysis.Analyzer.merge(Analyzer.java:512)
	at scala.tools.asm.tree.analysis.Analyzer.analyze(Analyzer.java:274)
	at scala.tools.nsc.backend.jvm.opt.CallGraph.analyzeCallsites(CallGraph.scala:121)
	at scala.tools.nsc.backend.jvm.opt.CallGraph.scala$tools$nsc$backend$jvm$opt$CallGraph$$$anonfun$1(CallGraph.scala:33)
	at scala.tools.nsc.backend.jvm.opt.CallGraph$$Lambda$5394/581054626.apply(Unknown Source)
	at scala.collection.TraversableLike$class.scala$collection$TraversableLike$class$$$anonfun$3(TraversableLike.scala:245)
	at scala.collection.TraversableLike$class$$Lambda$25/2106620844.apply(Unknown Source)
	at scala.collection.Iterator$class.foreach(Iterator.scala:781)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1267)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
```

**Comment by retronym** [comment]
```
⚡ qscalac -d /tmp test/files/run/patmat-exprs.scala # bootstrapped under this patch
/code/scala on topic-trait-defaults-module
⚡ qscalac -Yopt:l:classpath -d /tmp test/files/run/patmat-exprs.scala
^C
```

To find out what partest was doing when it hung, I captured a memory dump, and looked at live instance of `BatchSourceFile` to view the contents with YourKit.

**Pull request closed without merging**